File: ios/CalarSaat/Models/PrayerModels.swift
import Foundation

struct PrayerDay: Codable, Equatable {
    let tarih: String   // "dd.MM.yyyy"
    let imsak: String   // "HH:MM"
    let gunes: String
    let ogle: String
    let ikindi: String
    let aksam: String
    let yatsi: String
}

struct CityOption: Codable, Equatable {
    let id: String
    let name: String
}

struct DistrictOption: Codable, Equatable {
    let id: String
    let name: String
}

struct SelectedLocation: Codable, Equatable {
    let sehirId: String
    let sehirAdi: String
    let ilceId: String
    let ilceAdi: String
}

struct PrayerBundle: Codable, Equatable {
    let sehirId: String
    let sehirAdi: String
    let ilceId: String
    let ilceAdi: String
    let days: [PrayerDay]
    let fetchedAtEpochMillis: Int64

    private static let dateFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateFormat = "dd.MM.yyyy"
        formatter.locale = Locale(identifier: "tr_TR")
        return formatter
    }()

    /// Mirrors Android's `todayOrClosest()`: prefers the day matching today's
    /// date string; otherwise falls back to the last (or first) day available.
    func todayOrClosest() -> (day: PrayerDay, isToday: Bool)? {
        guard !days.isEmpty else { return nil }
        let todayString = Self.dateFormatter.string(from: Date())
        if let match = days.first(where: { $0.tarih == todayString }) {
            return (match, true)
        }
        return (days.last ?? days[0], false)
    }
}

enum RefreshResult {
    case success(PrayerBundle)
    case failure(String)
}

struct ImsakTime {
    let hour: Int
    let minute: Int
}

File: app/src/main/java/com/metehanyl/calarsaat/prayer/PrayerModels.kt
package com.metehanyl.calarsaat.prayer

/** Tek bir günün vakitleri. [tarih] formatı "dd.MM.yyyy" (Diyanet API formatı). */
data class PrayerDay(
    val tarih: String,
    val imsak: String,
    val gunes: String,
    val ogle: String,
    val ikindi: String,
    val aksam: String,
    val yatsi: String
)

data class CityOption(val id: String, val name: String)

data class DistrictOption(val id: String, val name: String)

/** GPS'ten çözümlenmiş il/ilçe bilgisi. */
data class SelectedLocation(
    val sehirId: String,
    val sehirAdi: String,
    val ilceId: String,
    val ilceAdi: String
)

/** Sunucudan çekilip diskte saklanan vakit verisi paketi. */
data class PrayerBundle(
    val sehirId: String,
    val sehirAdi: String,
    val ilceId: String,
    val ilceAdi: String,
    val days: List<PrayerDay>,
    val fetchedAtEpochMillis: Long
) {
    fun findDay(dateStr: String): PrayerDay? = days.firstOrNull { it.tarih == dateStr }

    /** Bugünün tarihiyle eşleşen günü, yoksa elimizdeki en yakın günü döner. */
    fun todayOrClosest(): Pair<PrayerDay, Boolean> {
        val todayStr = todayDateString()
        val exact = findDay(todayStr)
        if (exact != null) return exact to true
        val fallback = days.lastOrNull() ?: days.first()
        return fallback to false
    }

    companion object {
        fun todayDateString(): String =
            java.text.SimpleDateFormat("dd.MM.yyyy", java.util.Locale("tr", "TR")).format(java.util.Date())
    }
}

sealed class RefreshResult {
    data class Success(val bundle: PrayerBundle) : RefreshResult()
    data class Failure(val message: String) : RefreshResult()
}
